src/workerd/api/queue.c++

```diff
@@ -1,36 +1,114 @@
 #include "queue.h"
+#include "src/workerd/jsg/jsg.h"
 
 #include <workerd/jsg/ser.h>
 #include <workerd/api/global-scope.h>
 #include <kj/encoding.h>
 
 namespace workerd::api {
 
-kj::Promise<void> WorkerQueue::send(
-    v8::Local<v8::Value> body, jsg::Optional<SendOptions> options, v8::Isolate* isolate) {
-  auto& context = IoContext::current();
+kj::StringPtr validateContentType(kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    return IncomingQueueMessage::ContentType::TEXT;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::OCTET_STREAM) {
+    return IncomingQueueMessage::ContentType::OCTET_STREAM;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::JSON) {
+    return IncomingQueueMessage::ContentType::JSON;
+  }
+  if (contentType == IncomingQueueMessage::ContentType::V8) {
+    return IncomingQueueMessage::ContentType::V8;
+  }
 
-  JSG_REQUIRE(!body->IsUndefined(), TypeError, "Message body cannot be undefined");
+  JSG_FAIL_REQUIRE(TypeError, kj::str("Unsupported queue message content type: ", contentType));
+}
 
+kj::Array<kj::byte> serializeV8(jsg::Lock& js, v8::Local<v8::Value> body) {
   // Use a specific serialization version to avoid sending messages using a new version before all
   // runtimes at the edge know how to read it.
-  jsg::Serializer serializer(isolate, jsg::Serializer::Options {
+  jsg::Serializer serializer(js.v8Isolate, jsg::Serializer::Options {
     .version = 15,
     .omitHeader = false,
   });
   serializer.write(body);
-  kj::Array<kj::byte> serialized = serializer.release().data;
+  return serializer.release().data;
+}
 
-  auto client = context.getHttpClient(subrequestChannel, true, nullptr, "queue_send"_kj);
+kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::StringPtr contentType) {
+  if (contentType == IncomingQueueMessage::ContentType::TEXT) {
+    JSG_REQUIRE(
+      body->IsString(),
+      TypeError,
+      kj::str(
+        "Content Type \"",
+        IncomingQueueMessage::ContentType::TEXT,
+        "\" requires a value of type string, but received: ",
+        body->TypeOf(js.v8Isolate)
+      )
+    );
+
+    kj::String s = js.toString(body);
+    return kj::heapArray(s.asBytes());
```

Comment:
So, I tried this, along with all of the other optimizations you suggested, but struggled with various problems (e.g. the null terminator rears it's head in a few spots).

If you don't mind, maybe we could spend an hour or so pairing on this tomorrow or Friday, and we can try to see if we can make any progress here. I could also use some help with the exception cases (invalid content types or malformed data)

Reply:
Huh, I'm sorry for giving bad advice :(

I'm happy to do some pairing either of the next couple days, sorry I wasn't in the office today if you were there.

Reply:
Ugh, it looks like the problem is that they aren't equivalent, because `asBytes()` *doesn't* include the null terminator, and it seems that including it upsets the HTTP stack in some way.

This feels very fixable, but it's probably a better use of time to just throw in a `TODO(perf)` comment about it and move on until we're actually at a scale where it matters.

Reply:
Actually, I take that back. The HTTP stack is fine, it was just the .ew-test machinery complaining about the requests being made by this code no longer matching the expectations in your tests. If I make the change mentioned here and make corresponding changes to the ew-test expectations, it appears to work fine.

That isn't to say that we need to do it, but it does seem to work.

Reply:
Discussed offline - including the null terminator could be awkward when we start allowing users to push strings to us, since we wouldn't want to have to require them to include the null terminator. We could automatically add it in the queue broker in such cases, but that doesn't sound amazing either.

So we'll stick with not including the null terminator on the wire, and coming back to optimize this code later. The send side should still be able to avoid the copy by doing what I suggest here but also just slicing off the last byte (slightly more easily said than done but still very doable). The consumer side may have a tougher time, but in theory given that we're already having to copy the string from kj into v8, we could plausibly do the adding of the null terminator when doing the copy into v8 (unlike in this version of the code where we first do a copy within kj-land and then do a second copy into v8).